Funnel peerAdd and peerDelete in a channel

fcrisciani opened the pull request:
Remove the need for the wait group and avoid new
locks

Signed-off-by: Flavio Crisciani <flavio.crisciani@docker.com>

mavenugo commented on common/caller.go:
```diff
@@ -0,0 +1,29 @@
+
+func callerInfo(i int) (string, string, int, bool) {
+	ptr, file, line, ok := runtime.Caller(i)
+	var funName string
```
nit : "fun"Name ? :) .... is fName good enough ?

fcrisciani replied:
ok, no fun in production code :P

mavenugo commented on common/caller_test.go:
```diff
@@ -0,0 +1,49 @@
+	funName, _, _, ok = fun5()
+	if !ok || funName != "fun5" {
+		t.Fatalf("error on fun5 caller %t, %s", ok, funName)
+	}
```
These are certainly fun tests.. hence am okay for the `fun`%d function names ;)

fcrisciani commented:
Addressed the review comments, submitted the new patch, but let's wait to merge, I will test it locally to be safe, will post here once I've done that

fcrisciani commented:
@mavenugo ok tested the following way:
1) vendored this patch on latest moby master and build the binaries
2) created a 3 node cluster (1 manager, 2 workers)
3) created 100 services with 1 replica each
4) validate for each container:
    a) nslookup tasks count for each service
    b) number of ipvs rules in each namespace
    c) icmp to all the services
The result is PASS, grep also for error messages from the peer go routine, did not find any

abhinandanpb commented on common/caller.go:
```diff
@@ -0,0 +1,29 @@
+
+// CallerName returns the name of the function at the specified level with additional information
+// level == 0 means current method name
+func CallerName(level int) (string, string, int, bool) {
```
can this just return the function name. I don't think we will be using file and line number at all. We can use this directly in other debugs we have. I have closed the context PR docker/libnetwork#1835 because we can use this instead.

fcrisciani replied:
CallerName is not coming from the runtime, but is simply using some of the runtime functions, so I should be verified that the method results respects the method description

abhinandanpb commented on common/caller_test.go:
```diff
@@ -0,0 +1,49 @@
+package common
```
IMO I don't think we need tests for this because runtime.Caller() and runtime.FuncForPC() is part of runtime package. I don't think we should test the runtime package in libnetwork library.

fcrisciani replied:
CallerName is not coming from the runtime, but is simply using some of the runtime functions, so I should be verified that the method results respects the method description

thaJeztah commented:
back ported to 17.06: docker/libnetwork#1869
